vcd/resource_vcd_security_tag_test.go

```diff
@@ -0,0 +1,191 @@
+//go:build network || nsxt || ALL || functional
+// +build network nsxt ALL functional
+
+package vcd
+
+import (
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+	"strings"
+	"testing"
+)
+
+func TestAccVcdSecurityTag(t *testing.T) {
+	tag1 := strings.ToLower(t.Name() + "-tag1") // security tags are always lowercase in serverside
+	tag2 := strings.ToLower(t.Name() + "-tag2")
+
+	var params = StringMap{
+		"Org":          testConfig.VCD.Org,
+		"Vdc":          testConfig.Nsxt.Vdc,
+		"VappName":     t.Name() + "-vapp",
+		"VmName":       t.Name() + "-vm",
+		"ComputerName": t.Name() + "-vm",
+		"Catalog":      testConfig.VCD.Catalog.Name,
+		"Media":        testConfig.Media.MediaName,
+		"SecurityTag1": tag1,
+		"SecurityTag2": tag2,
+		"FuncName":     t.Name(),
+	}
+
+	if vcdShortTest {
+		t.Skip(acceptanceTestsSkipped)
+		return
+	}
+
+	vcdClient := createTemporaryVCDConnection(false)
+	if vcdClient.Client.APIVCDMaxVersionIs("< 36.0") {
+		t.Skip(t.Name() + " requires at least API v36.0 (VCD 10.3+)")
+	}
+
+	configText := templateFill(testAccSecurityTag, params)
+
+	params["FuncName"] = t.Name() + "-update"
+	configTextUpdate := templateFill(testAccSecurityTagUpdate, params)
+
+	debugPrintf("#[DEBUG] CONFIGURATION: %s\n", configText)
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: testAccProviders,
+		CheckDestroy:      testAccCheckSecurityTagDestroy(tag1),
+		Steps: []resource.TestStep{
+			{
+				Config: configText,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckSecurityTagCreated(tag1),
+				),
+			},
+			{
+				Config: configTextUpdate,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckSecurityTagCreated(tag1, tag2),
+				),
+			},
+			{
+				ResourceName:      fmt.Sprintf("vcd_security_tag.%s", tag1),
+				ImportState:       true,
+				ImportStateVerify: true,
+				ImportStateId:     testConfig.VCD.Org + "." + tag1,
+			},
+		},
+	})
+	postTestChecks(t)
+}
+
+// Templates to use on different stages
+const testAccSecurityTag = `
+resource "vcd_vapp" "{{.VappName}}" {
+  name = "{{.VappName}}"
+  org  = "{{.Org}}"
+  vdc  = "{{.Vdc}}"
+}
+
+resource "vcd_vapp_vm" "{{.VmName}}" {
+  org  = "{{.Org}}"
+  vdc  = "{{.Vdc}}"
+  vapp_name     = vcd_vapp.{{.VappName}}.name
+  name          = "VmWithoutTemplate"
+  computer_name = "emptyVM"
+  memory        = 2048
+  cpus          = 2
+  cpu_cores     = 1
+
+  os_type          = "sles10_64Guest"
+  hardware_version = "vmx-14"
+  catalog_name     = "{{.Catalog}}"
+  boot_image       = "{{.Media}}"
+}
+
+resource "vcd_security_tag" "{{.SecurityTag1}}" {
+  org    = "{{.Org}}"
+  name   = "{{.SecurityTag1}}"
+  vm_ids = [vcd_vapp_vm.{{.VmName}}.id]
```

Comment:
In the test below, I like how you test **two** security tags for the same VM, esp. as issues usually show up with >1 entities.

Similarly here, could we test this single SecurityTag1, but with **two** VMs in vm_ids?

Reply:
@lvirbalas I've updated the test to check that security tags can be applied to a several VMs at the same time.
Also I've written the additional function `testAccCheckSecurityTagOnVMCreated` which checks that a tag has been applied successfully to a VM.

Reply:
Perfect, thanks!